### DISCUSSION TITLE
Profiling Conv2d kernels by piggy-backing on Gemm profiler

### DISCUSSION THREAD

**Discussion by masahi** [NVIDIA/cutlass#358] [Q&A] [answered]
Hi, I'm going to add support for cutlass conv2d kernels in TVM. 

For gemm, we already have [a profiler ](https://github.com/apache/tvm/blob/main/python/tvm/contrib/cutlass/gemm_profiler.py) that compiles and runs many gemm kernels with different tile size etc to select the best one. I'm wondering if we should add a similar `conv2d_profiler`, or if we can piggy back on the gemm profiler with the implicit gemm shape `(N * H * W, OC, IC * KH * KW)` for `(M, N, K)` to profile and select tile sizes etc for conv2d kernels.

The latter solution is more desirable for me since it should simplify my implementation a lot. I used the same trick for adding batched gemm support. Is there a case where tile sizes etc selected for gemm with shape  `(N * H * W, OC, IC * KH * KW)` is not optimal for the equivalent conv2d? 

cc @manishucsd

**Comment by hwu36** [comment, marked as answer]
I think you may have to pick different ones for conv.  What do you think @manishucsd ?

**Reply by manishucsd** [reply]
Yes, winning conv2d kernel maybe different from gemm counter-part on (NPQ, K, RSC). 

I have never tried the second approach, but you can try it out on resnet50 and compare. The following post list some of the winning tile sizes for resnet50 layers with s4 input https://github.com/NVIDIA/cutlass/issues/154#issuecomment-747136529 with approach 1), i.e., running for all conv2d kernels for a layer.

**Reply by masahi** [reply]
ok thank you! I'll try the second approach for the initial implementation because it is easier to get things going, and see if the selected configs are different from the one chosen by the cutlass builtin profiler. If they turn out to be different, I'll add the conv2d profiler for TVM.
